packages/dds/tree/src/core/tree/visitDelta.ts

```diff
@@ -408,37 +408,50 @@ function detachPass(delta: Delta.FieldChanges, visitor: DeltaVisitor, config: Pa
 				visitNode(index, mark.fields, visitor, config);
 			}
 			if (isDetachMark(mark)) {
-				for (let i = 0; i < mark.count; i += 1) {
-					const root = config.detachedFieldIndex.createEntry(
-						// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-						offsetDetachId(mark.detach!, i),
-					);
-					if (mark.fields !== undefined) {
-						config.attachPassRoots.set(root, mark.fields);
-					}
-					const field = config.detachedFieldIndex.toFieldKey(root);
-					visitor.detach({ start: index, end: index + 1 }, field);
+				/* Create a single root for a contiguous range of detached nodes  */
+				const root = config.detachedFieldIndex.createEntry(mark.detach, mark.count);
+				if (mark.fields !== undefined) {
+					config.attachPassRoots.set(root, mark.fields);
 				}
+				const field = config.detachedFieldIndex.toFieldKey(root);
+				visitor.detach({ start: index, end: index + mark.count }, field);
 			} else if (!isAttachMark(mark)) {
 				index += mark.count;
 			}
 		}
 	}
 }
 
+/**
+ * For the purpose of non-atomization, the tree will not be built one-by-one on detached nodes.
+ * Instead, contiguous detached nodes are grouped into ranges, depending on whether the contiguous
+ * nodes have an identical root. Trees are then only built on nodes without a defined root.
+ */
 function buildTrees(
 	id: Delta.DetachedNodeId,
 	trees: readonly ITreeCursorSynchronous[],
 	config: PassConfig,
 	visitor: DeltaVisitor,
 ) {
-	for (let i = 0; i < trees.length; i += 1) {
-		const offsettedId = offsetDetachId(id, i);
-		let root = config.detachedFieldIndex.tryGetEntry(offsettedId);
-		assert(root === undefined, "Unable to build tree that already exists");
-		root = config.detachedFieldIndex.createEntry(offsettedId);
+	const nodeRangeToBuildTrees = config.detachedFieldIndex.getAllDetachedNodeRanges(
+		id,
+		trees.length,
+	);
+
+	if (nodeRangeToBuildTrees.length === 0) {
```

Comment:
while `nodeRangeToBuildTrees.length === 0` does mean that there are no entries for this range, that is not the only case: `nodeRangeToBuildTrees.length === 1 && nodeRangeToBuildTrees[0].root === undefined` also represents the absence of entries for that range, so the condition here would need to be `nodeRangeToBuildTrees.length === 0 || (nodeRangeToBuildTrees.length === 1 && nodeRangeToBuildTrees[0].root === undefined)`. In fact, even that is only correct because of the way `getAllDetachedNodeRanges` is implemented. If we only considered the return type of `getAllDetachedNodeRanges` and it's possible it could return an array containing an arbitrary large number of ranges with undefined root values. To get around all that, it's best to have a function that just returns the number of items in the range map for a given range, and check that number.